compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -1360,67 +1340,85 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
 
         let opaque_ty_def_id = self.local_def_id(opaque_ty_node_id);
 
-        let mut collected_lifetimes = FxHashMap::default();
+        let mut collected_lifetimes = Vec::new();
+        let mut new_remapping = FxHashMap::default();
+
         self.with_hir_id_owner(opaque_ty_node_id, |lctx| {
-            let hir_bounds = if origin == hir::OpaqueTyOrigin::TyAlias {
-                lower_bounds(lctx)
-            } else {
-                lctx.while_capturing_lifetimes(
+            if origin != hir::OpaqueTyOrigin::TyAlias {
+                let lifetimes_in_bounds =
```

Comment:
It's a nit, for now, but I think I would move this code:

```rust
let lifetimes_in_bounds = lifetime_collector::lifetimes_in_bounds(&lctx.resolver, bounds);
```

outside of the `with_hir_id_owner` call. This is because, conceptually to me, the visit takes place *before* we've "entered in" to the TAIT. It does mean we have to do something like:

```rust
// If this came from a TAIT (as opposed to an async function), we only
// want to capture the lifetimes that appear in the bounds. So visit the bounds
// to find out exactly which ones those are.
let lifetimes_to_remap = if origin == Alias {
    // in a TAIT like `type Foo<'a> = impl Foo<'a>`, we don't keep all the lifetime parameters
    vec![]
} else {
    // in fn return position, like the `fn test<'a>() -> impl Debug + 'a` example,
    // we only keep the lifetimes that appear in the `impl Debug` itself:
    lifetime_collector::lifetimes_in_bounds(...))
};
```

and then later

```rust
self.with_hir_id_owner(.., |lctx| {
    // If this opaque type is only capturing a subset of
    // the lifetimes (those that appear in bounds), then
    // create the new lifetime parameters required
    // and create a mapping from the old `'a` (on the function)
    // to the new `'a` (on the opaque type).
    collected_lifetimes = lctx.create_lifetime_defs(
                    opaque_ty_def_id,
                    &lifetimes_to_remap,
                    &mut new_remapping,
    );

    // Install the remapping from old to new (if any):
   lctx.with_remapping(new_remapping, |lctx| { ... });
```